### ISSUE TITLE
损失函数

### ISSUE THREAD

**Issue by horse-dog** [Phoenix-Shen/DSEN2_CR_PYTORCH#5]
carl_loss中，t.abs(predicted - input_cloudy)，看代码这一部分恒为0，不知道有何含义

**Comment by Phoenix-Shen** [comment]
刚刚看了下原文，这里确实是我搞错了，现在已经更正了，在`carl_loss`函数中，应该提供四个参数，这是修改之后的代码：

```python
# 这里多加了个参数input_cloudy
def carl_error(y_true,csm, y_pred,input_cloudy):
    """Computes the Cloud-Adaptive Regularized Loss (CARL)"""
    
    clearmask = t.ones_like(csm) - csm
    predicted = y_pred
    # input_cloudy = y_pred
    target = y_true

    cscmae = t.mean(clearmask * t.abs(predicted - input_cloudy) + csm * t.abs(
        predicted - target)) + 1.0 * t.mean(t.abs(predicted - target))
    
    return cscmae
```

同时 需要将train.py第100行中调用该函数的部分改成：
```python
loss= CARL_Loss(img_truth,img_csm,img_fake,img_cld)
```

我身边没有相关条件运行代码了，还请看看修改后的代码能否运行。

**Comment by horse-dog** [comment]
已、感谢作者

**Comment by lizijue** [comment]
> 刚刚看了下原文，这里确实是我搞错了，现在已经更正了，在`carl_loss`函数中，应该提供四个参数，这是修改之后的代码：
> 
> ```python
> # 这里多加了个参数input_cloudy
> def carl_error(y_true,csm, y_pred,input_cloudy):
>     """Computes the Cloud-Adaptive Regularized Loss (CARL)"""
>     
>     clearmask = t.ones_like(csm) - csm
>     predicted = y_pred
>     # input_cloudy = y_pred
>     target = y_true
> 
>     cscmae = t.mean(clearmask * t.abs(predicted - input_cloudy) + csm * t.abs(
>         predicted - target)) + 1.0 * t.mean(t.abs(predicted - target))
>     
>     return cscmae
> ```
> 
> 同时 需要将train.py第100行中调用该函数的部分改成：
> 
> ```python
> loss= CARL_Loss(img_truth,img_csm,img_fake,img_cld)
> ```
> 
> 我身边没有相关条件运行代码了，还请看看修改后的代码能否运行。

作者您好，在"loss= CARL_Loss(img_truth,img_csm,img_fake,img_cld)"这个语句中，img_cld应该是输入的光学图像，而非输入的SAR+光学的通道合并吧，但我联系上下文发现在“img_fake = net(img_cld)”中，img_cld就是两种模态的拼接，而您在计算CARL_Loss之前并没有把它们拆开，所以不知道这里是您写错了还是我理解错了？

**Comment by Phoenix-Shen** [comment]
> > 刚刚看了下原文，这里确实是我搞错了，现在已经更正了，在`carl_loss`函数中，应该提供四个参数，这是修改之后的代码：
> > ```python
> > # 这里多加了个参数input_cloudy
> > def carl_error(y_true,csm, y_pred,input_cloudy):
> >     """Computes the Cloud-Adaptive Regularized Loss (CARL)"""
> >     
> >     clearmask = t.ones_like(csm) - csm
> >     predicted = y_pred
> >     # input_cloudy = y_pred
> >     target = y_true
> > 
> >     cscmae = t.mean(clearmask * t.abs(predicted - input_cloudy) + csm * t.abs(
> >         predicted - target)) + 1.0 * t.mean(t.abs(predicted - target))
> >     
> >     return cscmae
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 同时 需要将train.py第100行中调用该函数的部分改成：
> > ```python
> > loss= CARL_Loss(img_truth,img_csm,img_fake,img_cld)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > 我身边没有相关条件运行代码了，还请看看修改后的代码能否运行。
> 
> 作者您好，在"loss= CARL_Loss(img_truth,img_csm,img_fake,img_cld)"这个语句中，img_cld应该是输入的光学图像，而非输入的SAR+光学的通道合并吧，但我联系上下文发现在“img_fake = net(img_cld)”中，img_cld就是两种模态的拼接，而您在计算CARL_Loss之前并没有把它们拆开，所以不知道这里是您写错了还是我理解错了？

不好意思，最近诸多琐事没来得及看。从文中的notation中来看$\mathbf{I}$是input OPTICAL image，所以你是对的，应该拆开，这里是我不仔细了。

另外作者似乎已经开源了他自己的代码，你也可以查看https://github.com/ameraner/dsen2-cr。
